source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -1189,6 +1126,149 @@ bool ContextImpl::verifyCertChain(X509& leaf_cert, STACK_OF(X509) & intermediate
+                                       "pairwise consistency test required in FIPS mode",
+                                       dataPath));
+    }
+  } break;
+  case EVP_PKEY_RSA: {
+    RSA* rsa_private_key = EVP_PKEY_get0_RSA(pkey.get());
+    if (!RSA_check_fips(rsa_private_key)) {
+      throw EnvoyException(fmt::format("Failed to load private key from {}, RSA key failed "
+                                       "pairwise consistency test required in FIPS mode",
+                                       dataPath));
+    }
+  } break;
+  }
+#endif
+}
+
+void TlsContext::loadPkcs12(const std::string& data, const std::string& dataPath,
+                            const std::string& password) {
+  cert_chain_file_path_ = dataPath;
+  bssl::UniquePtr<BIO> bio(BIO_new_mem_buf(const_cast<char*>(data.data()), data.size()));
+  RELEASE_ASSERT(bio != nullptr, "");
+  bssl::UniquePtr<PKCS12> pkcs12(d2i_PKCS12_bio(bio.get(), nullptr));
+
+  EVP_PKEY* temp_private_key = nullptr;
+  X509* temp_cert = nullptr;
+  STACK_OF(X509)* ca_certificates = nullptr;
+  if (pkcs12 == nullptr ||
+      !PKCS12_parse(pkcs12.get(), !password.empty() ? const_cast<char*>(password.c_str()) : nullptr,
+                    &temp_private_key, &temp_cert, &ca_certificates)) {
+    while (uint64_t err = ERR_get_error()) {
+      ENVOY_LOG_MISC(debug, "SSL error: {}:{}:{}:{}", err,
+                     absl::NullSafeStringView(ERR_lib_error_string(err)),
+                     absl::NullSafeStringView(ERR_func_error_string(err)), ERR_GET_REASON(err),
+                     absl::NullSafeStringView(ERR_reason_error_string(err)));
+    }
+    throw EnvoyException(absl::StrCat("Failed to load pkcs12 from ", dataPath));
+  }
+  if (ca_certificates) {
+    X509* ca_cert = nullptr;
+    while ((ca_cert = sk_X509_pop(ca_certificates)) != nullptr) {
+      //
+      // This transfers ownership to ssl_ctx therefore ca_cert does not need to be freed.
+      //
+      SSL_CTX_add_extra_chain_cert(ssl_ctx_.get(), ca_cert);
+    }
+  }
+  cert_chain_.reset(temp_cert);
+  bssl::UniquePtr<EVP_PKEY> pkey(temp_private_key);
+  if (!SSL_CTX_use_certificate(ssl_ctx_.get(), cert_chain_.get())) {
+    while (uint64_t err = ERR_get_error()) {
+      ENVOY_LOG_MISC(debug, "SSL error: {}:{}:{}:{}", err,
+                     absl::NullSafeStringView(ERR_lib_error_string(err)),
+                     absl::NullSafeStringView(ERR_func_error_string(err)), ERR_GET_REASON(err),
+                     absl::NullSafeStringView(ERR_reason_error_string(err)));
+    }
+    throw EnvoyException(absl::StrCat("Failed to load certificate from ", dataPath));
+  }
+  if (temp_private_key == nullptr || !SSL_CTX_use_PrivateKey(ssl_ctx_.get(), pkey.get())) {
+    throw EnvoyException(fmt::format("Failed to load private key from {}, Cause: {}", dataPath,
+                                     Utility::getLastCryptoError().value_or("unknown")));
+  }
+
+#ifdef BORINGSSL_FIPS
```

Comment:
is this branch tested?

Reply:
This branch appears untested for `private_key` and I haven't written a test for it for `pkcs12`.  I could make a common code path, since this is just copied from the `private_key` validation. But I'm not sure how to make a test for this.

Reply:
I see, I am wondering if it is expected that these branches remain untested. If so kinda scary.

@ggreenway how do we test these branches?

Reply:
I took a brief look and I don't see any test coverage for this.

I'm not familiar with all the intricacies of FIPS mode; I'm not sure how one would create a key that fails this test.

I agree that not having coverage of this isn't great, but it's not new in this PR, so I think it can be dealt with separately.

Reply:
It turns out it is just checking that the public and private keys match. A doc comment indicates this is required by `FIPS 140-2 4.9.2`. From BoringSSL:

```
int EC_KEY_check_fips(const EC_KEY *key) {
  if (EC_KEY_is_opaque(key)) {
    // Opaque keys can't be checked.
    OPENSSL_PUT_ERROR(EC, EC_R_PUBLIC_KEY_VALIDATION_FAILED);
    return 0;
  }

  if (!EC_KEY_check_key(key)) {
    return 0;
  }

  if (key->priv_key) {
    uint8_t data[16] = {0};
    ECDSA_SIG *sig = ECDSA_do_sign(data, sizeof(data), key);
#if defined(BORINGSSL_FIPS_BREAK_ECDSA_PWCT)
    data[0] = ~data[0];
#endif
    int ok = sig != NULL &&
             ECDSA_do_verify(data, sizeof(data), sig, key);
    ECDSA_SIG_free(sig);
    if (!ok) {
      OPENSSL_PUT_ERROR(EC, EC_R_PUBLIC_KEY_VALIDATION_FAILED);
      return 0;
    }
  }

  return 1;
}
```

Reply:
This is unrelated to this PR, since there is precedent already, but we have other branches as well that are `BORINGSSL_FIPS` specific.

Most of them are simple enough that you can reason about them in code review but I am worried about simple mistakes that might slip if these paths are hard to test. I assume that this code path is never exercised, not even in the local dev box, so it could be something as silly as missing a `;` that we might not notice.

Reply:
This code is run in CI in the compile-time-options build. There just isn't specific testing for the extra fips checks, but all the normal TLS tests are run, so most of this code is run, except the extra error paths.

Reply:
Good enough for me, I thought it was a complete blind spot. Thanks for explaining.